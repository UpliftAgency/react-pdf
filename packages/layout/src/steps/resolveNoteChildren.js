import * as R from 'ramda';

import isNote from '../node/isNote';
import isTextInstance from '../node/isTextInstance';

/**
 * Get node underlying text value
 *
 * @param {Object} node
 * @returns {String} node text content
 */
const getNodeText = node =>
  R.cond([
    [R.is(String), R.identity],
    [isTextInstance, R.prop('value')],
    [
      R.T,
      R.compose(
        getNodesText,
        R.propOr([], 'children'),
      ),
    ],
  ])(node);

/**
 * Get underlying text value of several nodes
 *
 * @param {Array} nodes
 * @returns {String} nodes text content
 */
const getNodesText = R.compose(
  R.join(''),
  R.map(getNodeText),
);

/**
 * Transforms string to text instance
 *
 * @param {String} value
 * @returns {Array} text intance
 */
const wrapTextInstance = value => [{ type: 'TEXT_INSTANCE', value }];

/**
 * Cast Note children as a text instance
 *
 * @param {Object} node
 * @returns {Object} node with resolved note children
 */
const resolveNoteChildren = node =>
  R.ifElse(
    isNote,
    R.evolve({
      children: R.compose(
        wrapTextInstance,
        getNodesText,
      ),
    }),
    R.evolve({ children: R.map(resolveNoteChildren) }),
  )(node);

export default resolveNoteChildren;
