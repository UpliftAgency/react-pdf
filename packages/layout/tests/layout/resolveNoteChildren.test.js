import resolveNoteChildren from '../../src/layout/resolveNoteChildren';

describe('layout resolveStyles', () => {
  test('should leave non note children as they are', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          children: [
            {
              type: 'VIEW',
              children: [
                {
                  type: 'TEXT',
                },
              ],
            },
          ],
        },
      ],
    };
    const result = resolveNoteChildren(root);

    expect(result).toMatchSnapshot();
  });

  test('should add note without children an empty text instance', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          children: [
            {
              type: 'NOTE',
              children: [],
            },
          ],
        },
      ],
    };
    const result = resolveNoteChildren(root);

    expect(result).toMatchSnapshot();
  });

  test('should leave note single note text instance children as it is', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          children: [
            {
              type: 'NOTE',
              children: [
                {
                  type: 'TEXT_INSTANCE',
                  value: 'React',
                },
              ],
            },
          ],
        },
      ],
    };
    const result = resolveNoteChildren(root);

    expect(result).toMatchSnapshot();
  });

  test('should concatenate two note text instance children into one', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          children: [
            {
              type: 'NOTE',
              children: [
                {
                  type: 'TEXT_INSTANCE',
                  value: 'React',
                },
                {
                  type: 'TEXT_INSTANCE',
                  value: '-pdf',
                },
              ],
            },
          ],
        },
      ],
    };
    const result = resolveNoteChildren(root);

    expect(result).toMatchSnapshot();
  });

  test('should concatenate several note children into one', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          children: [
            {
              type: 'NOTE',
              children: [
                {
                  type: 'TEXT_INSTANCE',
                  value: 'React ',
                },
                {
                  type: 'VIEW',
                  children: [
                    {
                      type: 'TEXT',
                      children: [
                        {
                          type: 'TEXT_INSTANCE',
                          value: 'freakin ',
                        },
                      ],
                    },
                    {
                      type: 'TEXT_INSTANCE',
                      value: 'pdf ',
                    },
                  ],
                },
                {
                  type: 'TEXT_INSTANCE',
                  value: 'ftw',
                },
              ],
            },
          ],
        },
      ],
    };
    const result = resolveNoteChildren(root);

    expect(result).toMatchSnapshot();
  });
});
