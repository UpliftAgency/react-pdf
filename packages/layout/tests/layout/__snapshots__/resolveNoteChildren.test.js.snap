// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout resolveStyles should add note without children an empty text instance 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "TEXT_INSTANCE",
              "value": "",
            },
          ],
          "type": "NOTE",
        },
      ],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles should concatenate several note children into one 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "TEXT_INSTANCE",
              "value": "React freakin pdf ftw",
            },
          ],
          "type": "NOTE",
        },
      ],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles should concatenate two note text instance children into one 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "TEXT_INSTANCE",
              "value": "React-pdf",
            },
          ],
          "type": "NOTE",
        },
      ],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles should leave non note children as they are 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "TEXT",
            },
          ],
          "type": "VIEW",
        },
      ],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles should leave note single note text instance children as it is 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "TEXT_INSTANCE",
              "value": "React",
            },
          ],
          "type": "NOTE",
        },
      ],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;
